pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java

```diff
@@ -0,0 +1,594 @@
+    byte[] dataSchemaBytes = null;
+    if (_dataSchema != null) {
+      dataSchemaBytes = _dataSchema.toBytes();
+      dataOutputStream.writeInt(dataSchemaBytes.length);
+      dataOffset += dataSchemaBytes.length;
+    } else {
+      dataOutputStream.writeInt(0);
+    }
+
+    // Write fixed size data section offset(START|SIZE).
+    dataOutputStream.writeInt(dataOffset);
+    if (_fixedSizeDataBytes != null) {
+      dataOutputStream.writeInt(_fixedSizeDataBytes.length);
+      dataOffset += _fixedSizeDataBytes.length;
+    } else {
+      dataOutputStream.writeInt(0);
+    }
+
+    // Write variable size data section offset(START|SIZE).
+    dataOutputStream.writeInt(dataOffset);
+    if (_variableSizeDataBytes != null) {
+      dataOutputStream.writeInt(_variableSizeDataBytes.length);
+      dataOffset += _variableSizeDataBytes.length;
+    } else {
+      dataOutputStream.writeInt(0);
+    }
+
+    // Write actual data.
+    // Write exceptions bytes.
+    dataOutputStream.write(exceptionsBytes);
+    // Write dictionary map bytes.
+    if (dictionaryMapBytes != null) {
+      dataOutputStream.write(dictionaryMapBytes);
+    }
+    // Write data schema bytes.
+    if (dataSchemaBytes != null) {
+      dataOutputStream.write(dataSchemaBytes);
+    }
+    // Write fixed size data bytes.
+    if (_fixedSizeDataBytes != null) {
+      dataOutputStream.write(_fixedSizeDataBytes);
+    }
+    // Write variable size data bytes.
+    if (_variableSizeDataBytes != null) {
+      dataOutputStream.write(_variableSizeDataBytes);
+    }
+
+    // Update the value of "executionThreadCpuTimeNs" to account data table serialization time.
+    long responseSerializationCpuTimeNs = threadTimer.stopAndGetThreadTimeNs();
+    long executionThreadCpuTimeNs =
+        Long.parseLong(getMetadata().getOrDefault(EXECUTION_THREAD_CPU_TIME_NS.getName(), "0")) + responseSerializationCpuTimeNs;
+    getMetadata().put(EXECUTION_THREAD_CPU_TIME_NS.getName(), String.valueOf(executionThreadCpuTimeNs));
+    // Copy all KV pair in _metadataV2 into _metadata
+    for (String key : _metadataV2.keySet()) {
+      Optional<MetadataKeys> opt = MetadataKeys.getByName(key);
+      if (!opt.isPresent()) {
+        continue;
+      }
+      _metadata.put(opt.get(), _metadataV2.get(key));
+    }
+    // Write metadata length and bytes.
+    byte[] metadataBytes = serializeMetadata();
+    dataOutputStream.writeInt(metadataBytes.length);
+    dataOutputStream.write(metadataBytes);
+
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  /**
+   * Serialize metadata section to bytes.
+   * Format of the bytes looks like:
+   * [numEntries, bytesOfKV2, bytesOfKV2, bytesOfKV3]
+   * For each KV pair:
+   * - if the value type is String, encode it as: [keyID, valueLength, Utf8EncodedValue].
+   * - if the value type is int, encode it as: [keyID, bigEndianRepresentationOfIntValue]
+   * - if the value type is long, encode it as: [keyID, bigEndianRepresentationOfLongValue]
```

Comment:
(nit) Add a comment here highlighting that unlike V2, numeric metadata values in V3 are not encoded in UTF-8 in the wire format. Instead big endian representation is used

Reply:
done